ferveo/src/lib.rs

```diff
@@ -145,17 +159,242 @@ mod test_dkg_full {
+                    &validator_keypair.decryption_key,
+                    validator_index,
+                    &updates_for_participant,
+                )
+            })
+            .collect();
+
+        // Now, we have to combine new share fragments into a new share
+        let new_private_key_share =
+            tpke::recover_share_from_updated_private_shares(
+                &x_r,
+                &domain_points,
+                &updated_shares,
+            );
+
+        // Get decryption shares from remaining participants
+        let mut decryption_shares: Vec<DecryptionShareSimple<E>> =
+            validator_keypairs
+                .iter()
+                .enumerate()
+                .map(|(validator_index, validator_keypair)| {
+                    pvss_aggregated.make_decryption_share_simple(
+                        &ciphertext,
+                        aad,
+                        &validator_keypair.decryption_key,
+                        validator_index,
+                        &dkg.pvss_params.g_inv(),
+                    )
+                })
+                .collect();
+
+        // Create a decryption share from a recovered private key share
+        let new_validator_decryption_key = Fr::rand(rng);
+        let validator_index = removed_validator.share_index;
+        decryption_shares.push(
+            DecryptionShareSimple::create(
+                validator_index,
+                &new_validator_decryption_key,
+                &new_private_key_share,
+                &ciphertext,
+                aad,
+                &dkg.pvss_params.g_inv(),
+            )
+            .unwrap(),
+        );
+
+        let lagrange = tpke::prepare_combine_simple::<E>(&domain_points);
+        let new_shared_secret =
+            tpke::share_combine_simple::<E>(&decryption_shares, &lagrange);
+
+        assert_eq!(old_shared_secret, new_shared_secret);
+    }
+
+    #[test]
+    fn simple_threshold_decryption_with_share_refreshing() {
+        let rng = &mut test_rng();
+        let dkg = setup_dealt_dkg_with_n_validators(3, 4);
+
+        let msg: &[u8] = "abc".as_bytes();
+        let aad: &[u8] = "my-aad".as_bytes();
+        let public_key = dkg.final_key();
+        let ciphertext = tpke::encrypt::<_, E>(msg, aad, &public_key, rng);
+
+        let validator_keypairs = gen_n_keypairs(4);
+        let pvss_aggregated = aggregate(&dkg);
+
+        // Create an initial shared secret
+        let (_, _, old_shared_secret) = make_shared_secret_simple_tdec(
+            &dkg,
+            aad,
+            &ciphertext,
+            &validator_keypairs,
+        );
+
+        // Now, we're going to refresh the shares and check that the shared secret is the same
+
+        // Dealer computes a new random polynomial with constant term x_r = 0
+        let polynomial = tpke::make_random_polynomial_at::<E>(
+            dkg.params.security_threshold as usize,
+            &Fr::zero(),
+            rng,
+        );
```

Comment:
This assumes we have a single dealer. Is there a test with N dealers and an aggregated update polynomial?

Reply:
There isn't one, but it's definitely worthwhile. I can add it in this or another PR.